### PR TITLE
provider/azurerm: Upgrading to v10.0.2 of the Azure SDK

### PR DESCRIPTION
**Pull request by tombuildsstuff**
~**Note: this is blocked on https://github.com/Azure/azure-sdk-for-go/issues/593**~

- [x] Core
  - [x] Vendor the Azure SDK for Go to v10.0.2
- [x] Azure Resource Manager Provider
  - [x] Upgrade the Provider to use the new SDK
  - [x] Acceptance Tests updated
  - [x] Acceptance Tests passing
  - [x] Switch to use the new Storage SDK
  - [x] Deprecate the Location fields in Event Hubs
  - [x] Container Registry - Update & Test
  - [x] Ensure documentation is up to date with ^
- [x] Azure Service Management Provider
  - [x] Upgrade the Provider to use the new SDK
  - [x] Acceptance Tests updated
  - [x] Acceptance Tests passing
  - [x] Switch to use the new Storage SDK
- [x] Azure Remote State
  - [x] Switch to use the new Storage SDK
  - [x] Acceptance Tests updated
  - [x] Acceptance Tests passing
- [x] Misc
  - [x] Rebase from Master

The SDK upgrade is needed so that we can proceed with #13634 / #12001

### CODE REVIEW

**Comment by dominik-lekse** [comment]
Is there currently a priority on upgrading to SDK v10 over merging in additional resources? If you need support, just let me know.

**Comment by tombuildsstuff** [comment]
Remote State tests pass:

```
$ envchain azurerm make testacc TEST=./state/remote TESTARGS='-run=TestAzure'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/05/19 11:29:58 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./state/remote -v -run=TestAzure -timeout 120m
=== RUN   TestAzureClient_impl
--- PASS: TestAzureClient_impl (0.00s)
=== RUN   TestAzureClient
2017/05/19 11:31:01 [DEBUG] New state was assigned lineage "270dc674-f10d-460c-88c9-3ed814354047"
2017/05/19 11:31:01 [TRACE] Preserving existing state lineage "270dc674-f10d-460c-88c9-3ed814354047"
--- PASS: TestAzureClient (155.51s)
=== RUN   TestAzureClientEmptyLease
2017/05/19 11:33:26 [DEBUG] New state was assigned lineage "e1ce5a88-5528-46f3-88c4-16fcbece2fab"
2017/05/19 11:33:26 [TRACE] Preserving existing state lineage "e1ce5a88-5528-46f3-88c4-16fcbece2fab"
--- PASS: TestAzureClientEmptyLease (144.84s)
=== RUN   TestAzureClientLease
2017/05/19 11:35:42 [DEBUG] New state was assigned lineage "cbcb6704-1703-47f2-9908-2e40b1f9fce6"
2017/05/19 11:35:42 [TRACE] Preserving existing state lineage "cbcb6704-1703-47f2-9908-2e40b1f9fce6"
--- PASS: TestAzureClientLease (136.27s)
PASS
ok  	github.com/hashicorp/terraform/state/remote	436.643s
```

**Comment by tombuildsstuff** [comment]
This is the result of the **Azure Service Management Provider** tests in this branch - where the outstanding failing tests are currently failing in `master`:

```
$ envchain azure make testacc TEST=./builtin/providers/azure TESTARGS='-run=TestAcc'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/05/19 15:32:47 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/azure -v -run=TestAcc -timeout 120m
=== RUN   TestAccAzureAffinityGroupBasic
--- PASS: TestAccAzureAffinityGroupBasic (19.29s)
=== RUN   TestAccAzureAffinityGroupUpdate
--- PASS: TestAccAzureAffinityGroupUpdate (29.74s)
=== RUN   TestAccAzureDataDisk_basic
--- PASS: TestAccAzureDataDisk_basic (294.31s)
=== RUN   TestAccAzureDataDisk_update
--- PASS: TestAccAzureDataDisk_update (632.54s)
=== RUN   TestAccAzureDnsServerBasic
--- PASS: TestAccAzureDnsServerBasic (25.58s)
=== RUN   TestAccAzureDnsServerUpdate
--- PASS: TestAccAzureDnsServerUpdate (54.80s)
=== RUN   TestAccAzureHostedServiceBasic
--- PASS: TestAccAzureHostedServiceBasic (17.61s)
=== RUN   TestAccAzureHostedServiceUpdate
--- PASS: TestAccAzureHostedServiceUpdate (52.85s)
=== RUN   TestAccAzureInstance_basic
--- PASS: TestAccAzureInstance_basic (177.55s)
=== RUN   TestAccAzureInstance_separateHostedService
--- PASS: TestAccAzureInstance_separateHostedService (207.06s)
=== RUN   TestAccAzureInstance_advanced
--- PASS: TestAccAzureInstance_advanced (366.63s)
=== RUN   TestAccAzureInstance_update
--- PASS: TestAccAzureInstance_update (660.67s)
=== RUN   TestAccAzureLocalNetworkConnectionBasic
--- PASS: TestAccAzureLocalNetworkConnectionBasic (21.40s)
=== RUN   TestAccAzureLocalNetworkConnectionUpdate
--- PASS: TestAccAzureLocalNetworkConnectionUpdate (63.90s)
=== RUN   TestAccAzureSecurityGroupRuleBasic
--- PASS: TestAccAzureSecurityGroupRuleBasic (152.18s)
=== RUN   TestAccAzureSecurityGroupRuleAdvanced
--- PASS: TestAccAzureSecurityGroupRuleAdvanced (132.19s)
=== RUN   TestAccAzureSecurityGroupRuleUpdate
--- PASS: TestAccAzureSecurityGroupRuleUpdate (303.76s)
=== RUN   TestAccAzureSecurityGroup_basic
--- PASS: TestAccAzureSecurityGroup_basic (77.90s)
=== RUN   TestAccAzureSqlDatabaseServerFirewallRuleBasic
--- PASS: TestAccAzureSqlDatabaseServerFirewallRuleBasic (41.37s)
=== RUN   TestAccAzureSqlDatabaseServerFirewallRuleAdvanced
--- FAIL: TestAccAzureSqlDatabaseServerFirewallRuleAdvanced (68.71s)
	testing.go:344: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Check failed: Error listing Azure Database Server Firewall Rules for Server "vc2qs7xcss": Error response from Azure. Code: , Message: Subscription '00000000-0000-0000-0000-000000000000' does not have the server 'vc2qs7xcss'.

		State: <no state>
=== RUN   TestAccAzureSqlDatabaseServerFirewallRuleUpdate
--- FAIL: TestAccAzureSqlDatabaseServerFirewallRuleUpdate (87.52s)
	testing.go:344: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Check failed: Error listing Azure Database Server Firewall Rules for Server "ftot322227": Error response from Azure. Code: ResourceNotFound, Message: The Uri doesn't match any specific service operation.

		State: <no state>
=== RUN   TestAccAzureSqlDatabaseServer
--- PASS: TestAccAzureSqlDatabaseServer (52.32s)
=== RUN   TestAccAzureSqlDatabaseServiceBasic
--- FAIL: TestAccAzureSqlDatabaseServiceBasic (160.30s)
	testing.go:344: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Check failed: Error issuing Azure SQL Service list request: Error response from Azure. Code: , Message: Subscription '00000000-0000-0000-0000-000000000000' does not have the server 'diyrupjlfb'.

		State: <no state>
=== RUN   TestAccAzureSqlDatabaseServiceAdvanced
--- FAIL: TestAccAzureSqlDatabaseServiceAdvanced (133.64s)
	testing.go:344: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Check failed: Error issuing Azure SQL Service list request: Error response from Azure. Code: , Message: Subscription '00000000-0000-0000-0000-000000000000' does not have the server 'r3g0ku7ssd'.

		State: <no state>
=== RUN   TestAccAzureSqlDatabaseServiceUpdate
--- FAIL: TestAccAzureSqlDatabaseServiceUpdate (203.06s)
	testing.go:344: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Check failed: Error issuing Azure SQL Service list request: Error response from Azure. Code: , Message: Subscription '00000000-0000-0000-0000-000000000000' does not have the server 'sziaazm58x'.

		State: <no state>
=== RUN   TestAccAzureStorageBlockBlob
--- PASS: TestAccAzureStorageBlockBlob (40.65s)
=== RUN   TestAccAzureStoragePageBlob
--- PASS: TestAccAzureStoragePageBlob (34.18s)
=== RUN   TestAccAzureStorageContainer
--- PASS: TestAccAzureStorageContainer (198.87s)
=== RUN   TestAccAzureStorageQueue
--- PASS: TestAccAzureStorageQueue (13.23s)
=== RUN   TestAccAzureStorageService
--- PASS: TestAccAzureStorageService (60.21s)
=== RUN   TestAccAzureVirtualNetwork_basic
--- PASS: TestAccAzureVirtualNetwork_basic (69.74s)
=== RUN   TestAccAzureVirtualNetwork_advanced
--- PASS: TestAccAzureVirtualNetwork_advanced (203.44s)
=== RUN   TestAccAzureVirtualNetwork_update
--- PASS: TestAccAzureVirtualNetwork_update (323.63s)
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/azure	4980.862s
make: *** [testacc] Error 1
```

This is the result of running (one of) the failing tests against the master branch ([once updating the SQL Server version to 12.0 as shown here](https://github.com/hashicorp/terraform/pull/14004/commits/ab0754e07fa7f317a44f7283757568f5e435a22c#diff-2ed940410ba07a7500c2589b96b22f5bR156)) - but they all produce the same output:
```
$ envchain azure make testacc TEST=./builtin/providers/azure TESTARGS='-run=TestAccAzureSqlDatabaseServiceBasic'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/05/19 15:23:10 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/azure -v -run=TestAccAzureSqlDatabaseServiceBasic -timeout 120m
=== RUN   TestAccAzureSqlDatabaseServiceBasic
--- FAIL: TestAccAzureSqlDatabaseServiceBasic (117.96s)
	testing.go:344: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Check failed: Error issuing Azure SQL Service list request: Error response from Azure. Code: , Message: Subscription '00000000-0000-0000-0000-000000000000' does not have the server 'mbctsv9asn'.

		State: <no state>
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/azure	117.978s
```

**Comment by tombuildsstuff** [comment]
@dominik-lekse sorry for the delay in responding to this comment - missed this!

> Is there currently a priority on upgrading to SDK v10 over merging in additional resources?

To provide some context here - unfortunately we were unable to upgrade to `9.0.X-beta` of the Azure SDK for Go last month due to some upstream issues (these issues have since been resolved in version `10.0.2-beta`).

Generally speaking these SDK upgrades can be a little painful, however we've never had one that's been delayed for this long - and unfortunately the changes required in upgrading to the new version of the SDK have also been larger than we anticipated, requiring changes in both of the Azure Providers (Service Management & Resource Manager) and the Azure Remote State. Unfortunately this has meant that we've been both blocking several open PR's (e.g. App Services) - as well as made things more complicated for new PR's coming in (given we're tied to an old version of the SDK).

There's no priority on upgrading the SDK to v10 over merging new resources - however being unable to upgrade to the latest SDK does put us into an unfortunate position, where every new resource which gets added has to be upgraded to use the new version of the SDK - and the longer we delay this PR, the more painful it gets.

In general we tend to roll the SDK upgrade out pretty quickly, however this taken longer than we anticipated, so apologies for any delays here. We're now in a position where the SDK upgrade is complete - so I'd expect this to be merged shortly :)

Hope that helps?

File: builtin/providers/azurerm/resource_arm_eventhub_namespace.go [hunk 1]
```diff
@@ -169,9 +170,15 @@ func resourceArmEventHubNamespaceDelete(d *schema.ResourceData, meta interface{}
 	resGroup := id.ResourceGroup
 	name := id.Path["namespaces"]
 
-	resp, err := namespaceClient.Delete(resGroup, name, make(chan struct{}))
+	delResp, error := namespaceClient.Delete(resGroup, name, make(chan struct{}))
+	resp := <-delResp
+	err = <-error
 
-	if resp.StatusCode != http.StatusNotFound {
+	if resp.StatusCode == http.StatusNotFound {
```
Comment (mbfrahry) [inline builtin/providers/azurerm/resource_arm_eventhub_namespace.go]: Do we want to return nil if the status is `StatusNotFound`?

Comment (tombuildsstuff) [inline builtin/providers/azurerm/resource_arm_eventhub_namespace.go]: Looking into this, internally the Azure SDK is polling for one of the status codes:

```
azure.WithErrorUnlessStatusCode(http.StatusNoContent, http.StatusOK, http.StatusAccepted)
```

As such, I've removed this if statement - as we can lean on the SDK to check for the Status Code errors here (which are returned as an error and will be handled below)

**Comment by tombuildsstuff** [comment]
Remote State tests still pass:

```
$ envchain azurerm make testacc TEST=./state/remote TESTARGS='-run=TestAzure'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/06/01 09:55:37 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./state/remote -v -run=TestAzure -timeout 120m
=== RUN   TestAzureClient_impl
--- PASS: TestAzureClient_impl (0.00s)
=== RUN   TestAzureClient
2017/06/01 09:56:25 [DEBUG] New state was assigned lineage "981f527d-d0be-4d3c-9013-27ba71af7e99"
2017/06/01 09:56:25 [TRACE] Preserving existing state lineage "981f527d-d0be-4d3c-9013-27ba71af7e99"
--- PASS: TestAzureClient (137.43s)
=== RUN   TestAzureClientEmptyLease
2017/06/01 09:58:40 [DEBUG] New state was assigned lineage "0524c612-c435-4618-983b-bc67efb27a74"
2017/06/01 09:58:40 [TRACE] Preserving existing state lineage "0524c612-c435-4618-983b-bc67efb27a74"
--- PASS: TestAzureClientEmptyLease (140.61s)
=== RUN   TestAzureClientLease
2017/06/01 10:01:01 [DEBUG] New state was assigned lineage "bba3c596-1304-407f-b56e-ca8187aa5176"
2017/06/01 10:01:01 [TRACE] Preserving existing state lineage "bba3c596-1304-407f-b56e-ca8187aa5176"
--- PASS: TestAzureClientLease (135.99s)
PASS
ok  	github.com/hashicorp/terraform/state/remote	414.059s
```

**Pull request merged**